test/integration/taproot.spec.ts

```diff
@@ -393,20 +416,25 @@ describe('bitcoinjs-lib (transaction with taproot)', () => {
       hash: unspent.txId,
       index: 0,
       witnessUtxo: { value: amount, script: output! },
-      witnessScript: redeem.output,
     });
+    psbt.updateInput(0, {
+      tapLeafScript: [
+        {
+          leafVersion: redeem.redeemVersion,
+          script: redeem.output,
+          controlBlock: witness![witness!.length - 1],
+        },
+      ],
+    });
+
     psbt.addOutput({ value: sendAmount, address: address! });
 
-    psbt.signInput(0, leafKeys[0]);
+    // random order for signers
     psbt.signInput(0, leafKeys[1]);
     psbt.signInput(0, leafKeys[2]);
+    psbt.signInput(0, leafKeys[0]);
 
-    const tapscriptFinalizer = buildTapscriptFinalizer(
-      internalKey.publicKey,
-      scriptTree,
-      regtest,
-    );
-    psbt.finalizeInput(0, tapscriptFinalizer);
+    psbt.finalizeInput(0);
```

Comment:
what do you think to add a test case where we need to add extra witness data on the stack to spend the input?

ie. a very dumb `OP_ADD 3 OP_EQUAL`
My understanding is that we need to pass custom finalizer function?

Reply:
Might be good to create a simple template, like 3-4 branches with OP_ADD {N} OP_EQUAL where N=3..6 or something.

Then have a few copy-pasted tests that try spending each of the paths.

Maybe the templates in csv and cltv integration tests can be of help? (maybe not)

Reply:
By popular demand, a taproot finalizer example has bee added. (see `...- custom finalizer ` test)
It build a tree with 8 leaves and then tries to spend them.

However, there is an issue with it and I could not figure out how to fix it.
When broadcasting the tx I get the error: `non-mandatory-script-verify-flag (Operation not valid with the current stack size)`
The final witness script looks correctly constructed:
![image](https://user-images.githubusercontent.com/2951406/170516146-c4667aee-ffb9-4d8d-bcb9-8312216466ce.png)

Script: `OP_ADD OP_2 OP_EQUAL`
Witness: `OP_1 OP_1`




Any ideas what I'm doing wrong?

Reply:
Can you paste the whole txhex that fails to broadcast in a copy-pasteable way?

Reply:
Sure, this is in [this link](https://bitcoincore.tech/apps/bitcoinjs-ui/index.html?tx_hex=02000000000101046c90993098a060bf9e1872067bcd0f12340cc902dda5b1095db1af8561ea910000000000ffffffff01904106000000000022512005a793464e0c3bd5f299a3712779f3ee2291d2d4246f830eb742343fc841a166030251510393528781c1bf8b8edaafa09256a30c1990bc822e506e79f74f01f1026c7f5041d26b13f5b1bbf3102f5a885ca9dc13fd5b6f54a40c409081b9702bae4ba4419fbcfe00a7b156491e8e3fd1f0051468b49dad9f9ccefd9e41b88b82c4aac491f967857f116999401a2485c859a8e2ab5f01570896db5efc9a7a98033be66f5216f4d96fba8000000000) url. You can copy the hex from the URL or use `Export`.

The partial/failed integration test is also committed on this branch.

Reply:
Ah, I see the issue -- you have the 2 OP_1s in a single witness element instead of 2 separate elements.

Reply:
I verified my solution works. I fixed the test and some test lints.